store-implementation/hbase-store/src/main/java/uk/gov/gchq/gaffer/hbasestore/operation/handler/AddElementsHandler.java

```diff
@@ -0,0 +1,143 @@
+            final String visibilityProperty = store.getSchema().getVisibilityProperty();
+            final Iterator<? extends Element> elements = addElementsOperation.getInput().iterator();
+            final ElementSerialisation serialisation = new ElementSerialisation(store.getSchema());
+            final int batchSize = store.getProperties().getWriteBufferSize();
+            final Map<Integer, Element> keyToElement = new HashMap<>(batchSize);
+            final Map<String, ElementAggregator> aggregators = new HashMap<>(store.getSchema().getEdges().size() + store.getSchema().getEntities().size());
+            while (elements.hasNext()) {
+                keyToElement.clear();
+                for (int i = 0; i < batchSize && elements.hasNext(); i++) {
+                    final Element element = elements.next();
+                    if (null == element) {
+                        i--;
+                        continue;
+                    }
+
+                    final int keyHashCode = new HashCodeBuilder(17, 37)
+                            .append(serialisation.getRowKeys(element).getFirst())
+                            .append(serialisation.getColumnVisibility(element))
+                            .append(serialisation.getColumnQualifier(element))
+                            .append(serialisation.getColumnVisibility(element))
+                            .toHashCode();
+                    final Element existingElement = keyToElement.get(keyHashCode);
+                    if (null == existingElement) {
+                        keyToElement.put(keyHashCode, element);
+                    } else if (hasAggregators) {
+                        ElementAggregator aggregator = aggregators.get(existingElement.getGroup());
+                        final SchemaElementDefinition elementDef = store.getSchema().getElement(existingElement.getGroup());
+                        if (null == aggregator) {
+                            aggregator = elementDef.getAggregator();
+                            aggregators.put(existingElement.getGroup(), aggregator);
+                        }
+                        Properties properties = element.getProperties();
+                        if (null != elementDef.getGroupBy() && !elementDef.getGroupBy().isEmpty()) {
+                            properties = properties.clone();
+                            properties.remove(elementDef.getGroupBy());
+                            properties.remove(visibilityProperty);
+                        }
+                        aggregator.apply(properties, existingElement.getProperties());
+                    } else {
+                        executePuts(table, createPuts(serialisation, keyToElement));
+                        keyToElement.clear();
+                        i = 0;
+                        keyToElement.put(keyHashCode, element);
+                    }
+                }
+
+                executePuts(table, createPuts(serialisation, keyToElement));
```

Comment:
Could this approach be replaced by:
- Using the MapStore to store and aggregate elements that need aggregating. Once the MapStore reaches a certain size the whole lot could be written to HBase.
- For elements that don't need aggregation, a dummy few bytes (say 4 bytes long and just a simple counter) could be added to the end of the column qualifier. This would ensure unique values. Whenever a key-value corresponding to a group that doesn't need aggregation is deserialised those bytes would need to be removed.